twinkle.js

```diff
@@ -467,6 +467,58 @@ Twinkle.makeFindSourcesDiv = function makeSourcesDiv(divID) {
 	}
 };
 
+Twinkle.notifyOnChanges = function notifyOnChanges() {
+	let checkIfChanges = function (response) {
+		if (response.query.pages[0].missing) {
+			mw.notify('This page has just been deleted!', {
+				title: 'Twinkle',
+				autoHide: false,
+				type: 'error'
+			});
+			return true;
+		}
+		if (response.query.pages[0].revisions) {
+			mw.notify('Page has been edited in the meantime. Reload to see latest changes.', {
+				title: 'Twinkle',
+				autoHide: false,
+				type: 'warn'
+			});
+			return true;
+		}
+		return false;
+	};
+	let api = new mw.Api();
+	api.get({
+		action: 'query',
+		format: 'json',
+		prop: 'revisions',
+		revids: mw.config.get('wgRevisionId'),
+		formatversion: '2',
+		rvprop: 'ids|timestamp',
+		rvslots: 'main'
+	}).then(function (response) {
+		if (!checkIfChanges(response)) {
+			let timestamp = response.query.pages[0].revisions[0].timestamp;
+			let interval = setInterval(function () {
+				api.get({
+					action: 'query',
+					prop: 'revisions',
+					titles: Morebits.pageNameNorm,
+					formatversion: '2',
+					rvprop: 'ids',
+					rvslots: 'main',
+					rvstart: new Date(new Date(timestamp).getTime() + 1000).toISOString(),
+					rvdir: 'newer'
+				}).then(function (response) {
+					if (checkIfChanges(response)) {
+						clearInterval(interval);
+					}
+				});
+			}, 5000);
```

Comment:
Will this run every 5 seconds that the page is open, indefinitely?

If the page loses focus, does this still run?

If so, this could really hammer the servers. Sometimes I'll keep dozens of tabs open overnight. If needed, we should consider writing a back-off algorithm. For example, change the check rate to every hour after the first 5 minutes. Or only run when the tab is focused.

Reply:
@siddharthvp , any thoughts on this one? I want to clear the PR queue this week :)

Reply:
This needs to be updated to stop polling once the Twinkle dialog has been closed. And we can include a document.hasFocus() check as well. I'll likely be able to get to this only next week :(

I think adding backoff is overcomplicating things, as people are presumably not going to keep Twinkle open for hours. Even if a few folks do, 1 read request every 5 seconds is nothing – the servers can handle a lot more than that. Besides, most browsers tend to throttle or suspend JS execution in inactive tabs.

Reply:
> 1 read request every 5 seconds is nothing – the servers can handle a lot more than that.

Some pages like watchlist and RecentChanges in core already do this kind of polling. So does DiscussionTools I think to show new comments on talk page? (not sure about DT as I use CD instead - which also does it btw).